### Questions  
1. What is the most watched movie in each year ?  
2. Which movie is the highest rated ?  
3. Which user watched the most movies ?
4. Top 10 list of movies with the highest number of comments

In [59]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

##### In this notebook, we will explore the combined_data_1.txt only, you can apply the same to other combined data set.

In [60]:
df = pd.read_csv('combined_data_1.txt', sep = ',', names = ['CID','Rating','Date'], parse_dates=['Date'])

In [61]:
df.head()

CID  Rating       Date
0       1:     NaN        NaT
1  1488844     3.0 2005-09-06
2   822109     5.0 2005-05-13
3   885013     4.0 2005-10-19
4    30878     4.0 2005-12-26

In [62]:
df.dtypes

CID               object
Rating           float64
Date      datetime64[ns]
dtype: object

* We will need 'Year' column for question 1

In [63]:
import datetime

In [64]:
df['Year'] = df['Date'].dt.year

In [65]:
df.head()

CID  Rating       Date    Year
0       1:     NaN        NaT     NaN
1  1488844     3.0 2005-09-06  2005.0
2   822109     5.0 2005-05-13  2005.0
3   885013     4.0 2005-10-19  2005.0
4    30878     4.0 2005-12-26  2005.0

* This is a real big data set with 24 milions rows

In [66]:
df.shape

(24058263, 4)

* If you take a good look at the dataset, you will see that, the first row contain "Movie ID" in "CID" column and NaN value on other columns  
then what below is all the rating for that movie, that mean,  
we need to remove these "Movie ID" row and create a new column for "Movie ID" that match with these rating from customers.

* Let's see where are these "Movie ID" located.

In [67]:
df_null = df[df.isnull().any(axis = 1)]
df_null

CID  Rating Date  Year
0            1:     NaN  NaT   NaN
548          2:     NaN  NaT   NaN
694          3:     NaN  NaT   NaN
2707         4:     NaN  NaT   NaN
2850         5:     NaN  NaT   NaN
...         ...     ...  ...   ...
24046714  4495:     NaN  NaT   NaN
24047329  4496:     NaN  NaT   NaN
24056849  4497:     NaN  NaT   NaN
24057564  4498:     NaN  NaT   NaN
24057834  4499:     NaN  NaT   NaN

[4499 rows x 4 columns]

* The DataFrame above showing that, rating for Movie with id = 1 start from index 1 to index 547,  
and movie id 2 from 547 to 693 and so on.  
So, we will create a new column, named "MID" stand for 'Movie ID',  
where row 1 to row 547 will have the value of 1, which match with the 'Movie ID' in df_null index 1.

So the algorith for this is: column 'MID' [row (1) to row (457)]  
        which is the same as: column 'MID' [df_null[index] to df_null[index]-1]

In [68]:
#for x in range(0,df_null.shape[0]-1):
    #df.loc[df_null.index[x]:df_null.index[x+1],('MID')] = x+1
"""The code should alike the one above but then in the last step, it would return out of index,
    So we need to add 1 row at the end as a placeholder so that it won't return error"""
a = {'CID':np.nan, 'Rating':np.nan,'Date':np.nan,'Year':np.nan}
df = df.append(a,ignore_index=True)

/var/folders/1s/mrnfdk9s7qnb7_cz4fxdkyrh0000gn/T/ipykernel_33363/1795572097.py:6: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(a,ignore_index=True)


update the df_null

In [69]:
df_null = df[df.isnull().any(axis = 1)]
df_null

CID  Rating Date  Year
0            1:     NaN  NaT   NaN
548          2:     NaN  NaT   NaN
694          3:     NaN  NaT   NaN
2707         4:     NaN  NaT   NaN
2850         5:     NaN  NaT   NaN
...         ...     ...  ...   ...
24047329  4496:     NaN  NaT   NaN
24056849  4497:     NaN  NaT   NaN
24057564  4498:     NaN  NaT   NaN
24057834  4499:     NaN  NaT   NaN
24058263    NaN     NaN  NaT   NaN

[4500 rows x 4 columns]

df_null has 4500 rows now

In [70]:
"""Create a new column and assign NaN to it as a placeholder"""
df['MID'] = np.nan

In [71]:
df.head()

CID  Rating       Date    Year  MID
0       1:     NaN        NaT     NaN  NaN
1  1488844     3.0 2005-09-06  2005.0  NaN
2   822109     5.0 2005-05-13  2005.0  NaN
3   885013     4.0 2005-10-19  2005.0  NaN
4    30878     4.0 2005-12-26  2005.0  NaN

In [72]:
df.tail()

CID  Rating       Date    Year  MID
24058259  1791000     2.0 2005-02-10  2005.0  NaN
24058260   512536     5.0 2005-07-27  2005.0  NaN
24058261   988963     3.0 2005-12-20  2005.0  NaN
24058262  1704416     3.0 2004-06-02  2004.0  NaN
24058263      NaN     NaN        NaT     NaN  NaN

In [73]:
for x in range(0,df_null.shape[0]-1):
    df.loc[df_null.index[x]+1:df_null.index[x+1],('MID')] = x+1

In [74]:
df

CID  Rating       Date    Year     MID
0              1:     NaN        NaT     NaN     NaN
1         1488844     3.0 2005-09-06  2005.0     1.0
2          822109     5.0 2005-05-13  2005.0     1.0
3          885013     4.0 2005-10-19  2005.0     1.0
4           30878     4.0 2005-12-26  2005.0     1.0
...           ...     ...        ...     ...     ...
24058259  1791000     2.0 2005-02-10  2005.0  4499.0
24058260   512536     5.0 2005-07-27  2005.0  4499.0
24058261   988963     3.0 2005-12-20  2005.0  4499.0
24058262  1704416     3.0 2004-06-02  2004.0  4499.0
24058263      NaN     NaN        NaT     NaN  4499.0

[24058264 rows x 5 columns]

* Then we drop all the NaN rows

In [79]:
df.dropna(subset = ['Rating'], axis = 0, inplace = True)

In [80]:
df.isnull().sum()

CID       0
Rating    0
Date      0
Year      0
MID       0
dtype: int64

There we go, cleaned dataset.

#### Question 1: What is the most watched movie in each year ?  

We group by Year, and count all the other values by Movie ID

In [81]:
df.groupby('Year')['MID'].value_counts()

Year    MID   
1999.0  295.0     23
        829.0     15
        4031.0    15
        4171.0    13
        1367.0     9
                  ..
2005.0  2343.0     5
        2593.0     5
        3867.0     5
        1984.0     3
        3100.0     2
Name: MID, Length: 16538, dtype: int64

* The table is quite long, since there is a lot of movie in 1 year.  
stick with the question 1, we only need to see which movie people were watched the most.  
so, the total of time that movie got rated = total of people watched the movie

In [76]:
df.groupby(['Year'])['MID'].apply(lambda x: x.value_counts().head(1))

Year          
1999.0  295.0         23
2000.0  985.0       3145
2001.0  571.0       8931
2002.0  571.0      16588
2003.0  4356.0     30890
2004.0  1905.0     89788
2005.0  3282.0    117015
Name: MID, dtype: int64

Or this, which is slightly quicker than code above

In [82]:
df.groupby('Year')['MID'].value_counts().groupby(level=0).head(1)

Year    MID   
1999.0  295.0         23
2000.0  985.0       3145
2001.0  571.0       8931
2002.0  571.0      16588
2003.0  4356.0     30890
2004.0  1905.0     89788
2005.0  3282.0    117015
Name: MID, dtype: int64

#### Question 2: Which movie is highest rated ?

We select the column MID and Rating, then group it by MID and calc the average of Rating then sort it.

In [93]:
df.loc[:,(['Rating','MID'])].groupby('MID').mean().sort_values('Rating')

Rating
MID             
515.0   1.287879
3575.0  1.400000
2426.0  1.452381
4202.0  1.487395
1227.0  1.496241
...          ...
13.0    4.552000
4238.0  4.554434
2102.0  4.581296
3033.0  4.586364
3456.0  4.670989

[4499 rows x 1 columns]

* So the movie with highest rated = 4.67 have MID = 3456

#### Question 3: Which user watched the most movies ?

We looking for most requent values in the CID column

In [96]:
df['CID'].value_counts()

305344     4467
387418     4422
2439493    4195
1664010    4019
2118461    3769
           ... 
1300341       1
2550360       1
11848         1
930788        1
594210        1
Name: CID, Length: 470758, dtype: int64

So we have customer with ID 305344, this person have watched 4467 movies

#### Question 4: Top 10 list movies with the highest number of comments.

i don't think we have the comment dataset there.